### PR TITLE
DS-4107 Metadata as map updates for Angular

### PR DESCRIPTION
**Pull request by cwilper**
Updates the angular code to consume and work with metadata modeled as a map. Some code consolidation was done here, so metadata-processing code is done in a common place, which is now well-covered by an included set of tests.

To work, this requires REST PR: https://github.com/DSpace/DSpace/pull/2287 which returns DSO metadata in the expected map form.

For more information, see https://jira.duraspace.org/browse/DS-4107

### CODE REVIEW

**Comment by cwilper** [comment]
Rebased on latest master. This is ready for review and when approved, should be merged at the same time as the corresponding REST implementation: https://github.com/DSpace/DSpace/pull/2287

File: src/app/core/shared/metadata.model.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  public static firstValue(mdMapOrMaps: MetadataMap | MetadataMap[], keyOrKeys: string | string[],
+                           filter?: MetadataValueFilter): string {
+    const value = Metadata.first(mdMapOrMaps, keyOrKeys, filter);
+    return value === undefined ? undefined : value.value;
+  }
+
+  /** Checks the given map for a matching value. */
+  public static has(mdMapOrMaps: MetadataMap | MetadataMap[], keyOrKeys: string | string[],
+                    filter?: MetadataValueFilter): boolean {
+    return Metadata.first(mdMapOrMaps, keyOrKeys, filter) !== undefined;
```
Comment (atarix83) [inline src/app/core/shared/metadata.model.ts]: you could use isNotUndefined util method

File: src/app/shared/object-grid/search-result-grid-element/search-result-grid-element.component.ts [hunk 2]
```diff
@@ -22,39 +21,14 @@ export class SearchResultGridElementComponent<T extends SearchResult<K>, K exten
...
-    );
-    if (isEmpty(results)) {
-      this.dso.filterMetadata(keys).forEach(
-        (md: Metadatum) => {
-          results.push(md.value);
-        }
-      );
-    }
-    return results;
+  /** Gets all matching metadata string values from hitHighlights or dso metadata, preferring hitHighlights. */
```
Comment (atarix83) [inline src/app/shared/object-grid/search-result-grid-element/search-result-grid-element.component.ts]: Please use Supported TypeDoc tags to comment this method

File: src/app/shared/object-grid/search-result-grid-element/search-result-grid-element.component.ts [hunk 3]
```diff
@@ -22,39 +21,14 @@ export class SearchResultGridElementComponent<T extends SearchResult<K>, K exten
...
-          result = md.value;
-          return true;
-        }
-      }
-    );
-    if (hasNoValue(result)) {
-      result = this.dso.findMetadata(key);
-    }
-    return result;
+  /** Gets the first matching metadata string value from hitHighlights or dso metadata, preferring hitHighlights. */
```
Comment (atarix83) [inline src/app/shared/object-grid/search-result-grid-element/search-result-grid-element.component.ts]: Please use Supported TypeDoc tags to comment this method

File: src/app/shared/object-list/search-result-list-element/search-result-list-element.component.ts [hunk 4]
```diff
@@ -22,39 +21,14 @@ export class SearchResultListElementComponent<T extends SearchResult<K>, K exten
...
-    );
-    if (isEmpty(results)) {
-      this.dso.filterMetadata(keys).forEach(
-        (md: Metadatum) => {
-          results.push(md.value);
-        }
-      );
-    }
-    return results;
+  /** Gets all matching metadata string values from hitHighlights or dso metadata, preferring hitHighlights. */
```
Comment (atarix83) [inline src/app/shared/object-list/search-result-list-element/search-result-list-element.component.ts]: Please use Supported TypeDoc tags to comment this method

File: src/app/shared/object-list/search-result-list-element/search-result-list-element.component.ts [hunk 5]
```diff
@@ -22,39 +21,14 @@ export class SearchResultListElementComponent<T extends SearchResult<K>, K exten
...
-          result = md.value;
-          return true;
-        }
-      }
-    );
-    if (hasNoValue(result)) {
-      result = this.dso.findMetadata(key);
-    }
-    return result;
+  /** Gets the first matching metadata string value from hitHighlights or dso metadata, preferring hitHighlights. */
```
Comment (atarix83) [inline src/app/shared/object-list/search-result-list-element/search-result-list-element.component.ts]: Please use Supported TypeDoc tags to comment this method

File: src/app/core/browse/browse.service.ts [hunk 6]
```diff
@@ -42,16 +42,16 @@ import { DSpaceObject } from '../shared/dspace-object.model';
 export class BrowseService {
```
[the remark this answers is not in the document]
Comment (cwilper) [reply src/app/core/browse/browse.service.ts]: Ok, I went ahead and added a basic class-level TypeDoc here, but notice that the change to this particular file was not a functional one...it was just renaming a variable for consistency with other touched code.

**Comment by cwilper** [comment]
Thanks for the the review @atarix83 . I have rebased, resolving conflicts, and improved TypeDocs and used isUndefined/isNotUndefined as suggested.

**Comment by cwilper** [comment]
Looks like checks have passed, so once @atarix83 approves, I believe this can be merged.

**Note: whoever merges this, please merge the REST impl side at the same time (already approved): https://github.com/DSpace/DSpace/pull/2287**

**Comment by atarix83** [comment]
@cwilper thanks for the changes.
It looks good to me now

File: src/app/+item-page/edit-item-page/modify-item-overview/modify-item-overview.component.html [hunk 7]
```diff
@@ -7,10 +7,12 @@
...
   <tbody>
-  <tr *ngFor="let metadatum of metadata" class="metadata-row">
-    <td>{{metadatum.key}}</td>
-    <td>{{metadatum.value}}</td>
-    <td>{{metadatum.language}}</td>
-  </tr>
+  <ng-container *ngFor="let entry of metadata | keyvalue">
+    <tr *ngFor="let value of entry.value" class="metadata-row">
+      <td>{{entry.key}}</td>
+      <td>{{value.value}}</td>
```
Comment (paulo-graca) [inline src/app/+item-page/edit-item-page/modify-item-overview/modify-item-overview.component.html]: I dislike having `value.value` I suggest to rename `entry `to `mdItem` and `value` to `mdValue`. Then you will have `mdValue.value`

File: src/app/+item-page/full/full-item-page.component.html [hunk 8]
```diff
@@ -9,11 +9,13 @@
       </div>
       <table class="table table-responsive table-striped">
         <tbody>
-        <tr *ngFor="let metadatum of (metadata$ | async)">
-          <td>{{metadatum.key}}</td>
-          <td>{{metadatum.value}}</td>
-          <td>{{metadatum.language}}</td>
-        </tr>
+          <ng-container *ngFor="let entry of (metadata$ | async) | keyvalue">
```
Comment (paulo-graca) [inline src/app/+item-page/full/full-item-page.component.html]: as I suggested, replacement of `entry` with `mdItem` and `value` with `mdValue`

**Comment by paulo-graca** [comment]
This PR doesn't work properly, metadata isn't showed. I'm using atmire's REST endpoints. Please check my print-screen (https://drive.google.com/open?id=1bJypXyDXECJP7BlJ3LXsTGtR45dX2PXz).

**Comment by cwilper** [comment]
@paulo-graca thanks for taking a look. I have adjusted variable naming to hopefully be more clear. I liked your suggestion about using `mdValue` and `mdValues` for clarity, and implemented that. You also suggested renaming `entry` to `mdItem`. I agree with making it more specific (prefix with md, consistent with the others), but I think "Entry" is really the right term here rather than "Item" -- it is a single _entry_ of a map, which contains a `key` and `value` property. So I called it `mdEntry`. Please let me know if that works ok for you.

> This PR doesn't work properly, metadata isn't showed.

This PR requires that you run with a specific REST PR or it won't work (see description). That PR is not yet running on Atmire's server and probably won't (permanently) until *both* PRs are merged. So I would suggest running a local copy of the the REST PR in order to test functionality.

**Comment by paulo-graca** [comment]
👍 @cwilper thank you for considering my suggestions! I will try to do a new test with the dependency DSpace/DSpace#2287

**Comment by tdonohue** [comment]
Already at +2, and it looks great to me too (did a quick scan of the code).  Merging.  Thanks @cwilper!

**Pull request merged**